indico/modules/events/registration/wallets/apple.py

```diff
@@ -0,0 +1,144 @@
+        reg_url = url_for('event_registration.display_regform', registration.locator.uuid, _external=True)
+        ticket.addBackField('back-registration-details', f'<a href="{reg_url}">{_("Registration details")}</a>',
+                            _('Link'))
+
+        signals.event.registration.apple_wallet_ticket_object.send(self.event, obj=ticket)
+        return ticket
+
+    def build_pass_object(self, registration):
+        from indico.modules.events.registration.util import get_persons, get_ticket_qr_code_data
+
+        # Extract Certificate details from certificate.pem itself
+        self.cert = x509.load_pem_x509_certificate(self.settings['apple_wallet_certificate'].encode())
+        cert_details = {d.split('=')[0]: d.split('=')[1] for d in self.cert.subject.rfc4514_string().split(',')}
+        passfile = IndicoPass(self.build_ticket_object(registration), passTypeIdentifier=cert_details['UID'],
+                              organizationName=cert_details['O'], teamIdentifier=cert_details['OU'])
+        passfile.logoText = cert_details['O']  # Add Organization name at the top
+        passfile.backgroundColor = '#007cac'
+        passfile.foregroundColor = passfile.labelColor = '#ffffff'
+        passfile.serialNumber = registration.ticket_uuid
+        qr_data = get_ticket_qr_code_data(get_persons([registration])[0])
+        passfile.barcode = Barcode(message=json.dumps(qr_data, separators=(',', ':')), format=BarcodeFormat.QR)
+
+        # Including the icon and logo is necessary for the passbook to be valid.
+        if config.WALLET_LOGO_URL:
+            passfile.add_file_from_url('icon.png', config.ABSOLUTE_WALLET_LOGO_URL)
+            passfile.add_file_from_url('logo.png', config.ABSOLUTE_WALLET_LOGO_URL)
+        else:
+            passfile.add_default_images()
+        signals.event.registration.apple_wallet_object.send(self.event, obj=passfile)
+        return passfile
+
+    def get_ticket(self, registration) -> str:
+        passfile = self.build_pass_object(registration)
+        wwdr_path = os.path.join(current_app.root_path, 'modules', 'events', 'registration', 'wallets',
+                                 'apple-wwdr.pem')
+        # Create and output the Passbook file (.pkpass)
+        temp = tempfile.NamedTemporaryFile(prefix='apple_wallet_', dir=config.TEMP_DIR, delete=False)
+        registration.apple_wallet_serial = passfile.serialNumber  # Save ticket serial for updates
+        try:
+            pkpass = passfile.create(self.cert, self.settings['apple_wallet_key'], wwdr_path,
+                                     self.settings['apple_wallet_password'],
+                                     zip_file=os.path.join(config.TEMP_DIR, temp.name))
+        except ValueError as exc:
+            logger.warning('Could not create Apple Pass ticket: %s', exc)
+            raise
+        return pkpass
+        # TBD: Add Celery task to delete up to 2 days ago generated files?
```

Comment:
Much easier option: Use a `BytesIO` instead of the `NamedTemporaryFile`. After creating the pkpass you just need to` .seek(0)` on it and then you can use that w/o ever saving something to the disk.

Reply:
Done.

Reply:
FYI, this did not work because you seeked to the beginning of an already empty IO (you'd do that *after* creating the zipfile inside it), so what you returned was a file object with the internal pointer at its end, so appearing empty for anything just trying to read it.

In any case, I fixed this and removed creating it manually, because `passfile.create()` already contains this logic internally:

```py
        if not zip_file:
            zip_file = BytesIO()
        self._createZip(pass_json, manifest, signature, zip_file=zip_file)
        return zip_file
```

This is exactly why I always tell people to look at the code of the library functions they're using if something doesn't work or seems unnecessarily complex.